x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosAuthenticationToken.java

```diff
@@ -0,0 +1,114 @@
+     * Extract token from header and if valid {@link #NEGOTIATE_AUTH_HEADER} then
+     * returns {@link KerberosAuthenticationToken}
+     *
+     * @param context {@link ThreadContext}
+     * @return returns {@code null} if {@link #AUTH_HEADER} is empty or not an
+     *         {@link #NEGOTIATE_AUTH_HEADER} else returns valid
+     *         {@link KerberosAuthenticationToken}
+     */
+    public static KerberosAuthenticationToken extractToken(final ThreadContext context) {
+        final String authHeader = context.getHeader(AUTH_HEADER);
+        if (Strings.isNullOrEmpty(authHeader)) {
+            return null;
+        }
+
+        if (authHeader.startsWith(NEGOTIATE_AUTH_HEADER) == Boolean.FALSE) {
+            return null;
+        }
+
+        final String base64EncodedToken = authHeader.substring(NEGOTIATE_AUTH_HEADER.length()).trim();
+        if (Strings.isEmpty(base64EncodedToken)) {
+            throw unauthorized("invalid negotiate authentication header value, expected base64 encoded token but value is empty", null);
+        }
+        final byte[] base64Token = base64EncodedToken.getBytes(StandardCharsets.UTF_8);
+        byte[] decodedKerberosTicket = null;
+        IllegalArgumentException rootCause = null;
+        try {
+            decodedKerberosTicket = Base64.getDecoder().decode(base64Token);
+        } catch (IllegalArgumentException iae) {
+            rootCause = iae;
+        }
+
+        if (decodedKerberosTicket == null || decodedKerberosTicket.length == 0) {
+            throw unauthorized("invalid negotiate authentication header value, could not decode base64 token {}", rootCause,
+                    base64EncodedToken);
+        }
+        return new KerberosAuthenticationToken(base64EncodedToken);
+    }
+
+    @Override
+    public String principal() {
+        return principalName;
+    }
+
+    @Override
+    public Object credentials() {
+        return base64EncodedTicket;
+    }
+
+    @Override
+    public void clearCredentials() {
+        this.base64EncodedTicket = null;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(principalName, base64EncodedTicket);
+    }
+
+    @Override
+    public boolean equals(final Object other) {
+        if (this == other)
+            return true;
+        if (other == null)
+            return false;
+        if (getClass() != other.getClass())
+            return false;
+        final KerberosAuthenticationToken otherKerbToken = (KerberosAuthenticationToken) other;
+        return Objects.equals(otherKerbToken.principal(), principal()) && Objects.equals(otherKerbToken.credentials(), credentials());
+    }
+
+    private static ElasticsearchSecurityException unauthorized(final String message, final Throwable cause, final Object... args) {
+        ElasticsearchSecurityException ese = new ElasticsearchSecurityException(message, RestStatus.UNAUTHORIZED, cause, args);
+        ese.addHeader(WWW_AUTHENTICATE, NEGOTIATE_AUTH_HEADER);
```

Comment:
Is this right? (It might be, I just want to know that we've thought about it).
If we receive an invalid Kerberos token, we request to negotiate again - are you sure that's what we want?